File: android-kotlin/QuickStartTasks/app/src/androidTest/java/live/ditto/quickstart/tasks/DittoManagerTests.kt
package live.ditto.quickstart.tasks

import android.util.Log
import androidx.test.platform.app.InstrumentationRegistry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import live.ditto.android.DefaultAndroidDittoDependencies
import live.ditto.quickstart.tasks.data.CustomDirectoryAndroidDittoDependencies
import live.ditto.quickstart.tasks.data.DittoManager
import live.ditto.quickstart.tasks.models.DittoConfig
import live.ditto.quickstart.tasks.data.TaskModel
import live.ditto.quickstart.tasks.services.ErrorService
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.util.UUID

/**
 * See [testing documentation](http://d.android.com/tools/testing).
 */
@OptIn(ExperimentalCoroutinesApi::class)
class DittoManagerTests {

    @Test
    fun useAppContext() {
        // Context of the app under test.
        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
        assertEquals("live.ditto.quickstart.tasks", appContext.packageName)
    }

    @Test
    fun testDittoPopulateTaskCollection() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val expectedTitles = listOf(
                "Buy groceries",
                "Clean the kitchen",
                "Schedule dentist appointment",
                "Pay bills"
            )

            //act
            dittoManager.populateTaskCollection()
            advanceUntilIdle()
            val results = dittoManager.ditto?.store?.execute("SELECT * FROM tasks WHERE NOT deleted")
            val tasks = results?.items?.map { TaskModel.fromMap(it.value) } ?: emptyList()


            //assert
            assertEquals(expectedTitles.size, tasks.size)
            assertEquals(
                expectedTitles.sorted(),
                tasks.map { it.title }.sorted()
            )
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    @Test
    fun testDittoInsertTaskModel() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val newTask = createInitialTaskModel()

            //act
            dittoManager.insertTaskModel(newTask)
            advanceUntilIdle()

            //assert
            val tasks = dittoManager.getTaskModels().first()
            val newCount = tasks.size
            val insertedTask = tasks.find { it._id == newTask._id }
            assertEquals(1, newCount)
            assertEquals(newTask, insertedTask)
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    @Test
    fun testDittoUpdateTaskModel() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val newTask = createInitialTaskModel()
            dittoManager.insertTaskModel(newTask)
            advanceUntilIdle()

            //act
            val updatedTask = newTask.copy(title = "Updated Task", done = true)
            dittoManager.updateTaskModel(updatedTask)
            advanceUntilIdle()

            //assert
            val tasks = dittoManager.getTaskModels().first()
            val newCount = tasks.size
            val insertedTask = tasks.find { it._id == updatedTask._id }
            assertEquals(1, newCount)
            assertNotNull(insertedTask)
            assertEquals(updatedTask, insertedTask)
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    @Test
    fun testDittoUpdateTaskModel_WhenDeleted() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val newTask = createInitialTaskModel()
            dittoManager.insertTaskModel(newTask)
            advanceUntilIdle()

            //act
            val updatedTask = newTask.copy(deleted = true)
            dittoManager.updateTaskModel(updatedTask)
            advanceUntilIdle()

            //assert
            val tasks = dittoManager.getTaskModels().first()
            val newCount = tasks.size
            assertEquals(0, newCount)
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    @Test
    fun testDittoToggleComplete() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val newTask = createInitialTaskModel()
            dittoManager.insertTaskModel(newTask)
            advanceUntilIdle()

            //act
            dittoManager.toggleComplete(newTask._id)
            advanceUntilIdle()

            //assert
            val tasks = dittoManager.getTaskModels().first()
            val newCount = tasks.size
            assertEquals(1, newCount)
            val insertedTask = tasks.find { it._id == newTask._id }
            assertNotNull(insertedTask)
            if (insertedTask != null) {
                assertTrue(insertedTask.done)
            }
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    @Test
    fun testDittoDeleteTaskModel() = runTest {
        val dittoManager = createDittoManagerForTests()
        try {
            //arrange
            val newTask = createInitialTaskModel()
            dittoManager.insertTaskModel(newTask)
            advanceUntilIdle()

            //act
            dittoManager.deleteTaskModel(newTask._id)
            advanceUntilIdle()

            //assert
            val tasks = dittoManager.getTaskModels().first()
            val newCount = tasks.size
            assertEquals(0, newCount)
        } finally {
            cleanUpCollection(dittoManager)
        }
    }

    private fun createInitialTaskModel() : TaskModel {
        return TaskModel(
            _id = UUID.randomUUID().toString(),
            title = "Test Task",
            done = false,
            deleted = false)
    }

    private suspend fun createDittoManagerForTests() : DittoManager {
        // Create a temporary directory for testing
        val tempDir = withContext(Dispatchers.IO) {
            File.createTempFile("ditto_test_${UUID.randomUUID()}", null)
        }
        tempDir.delete()
        tempDir.mkdirs()

        val appContext = InstrumentationRegistry.getInstrumentation().targetContext

        val dittoConfig = DittoConfig(
            authUrl = BuildConfig.DITTO_AUTH_URL,
            websocketUrl = "",
            appId = BuildConfig.DITTO_APP_ID,
            authToken = BuildConfig.DITTO_PLAYGROUND_TOKEN
        )

        val customDependencies = CustomDirectoryAndroidDittoDependencies(
            DefaultAndroidDittoDependencies(appContext),
            tempDir
        )

        val dittoManager = DittoManager(
            dittoConfig = dittoConfig,
            androidDittoDependencies = customDependencies,
            errorService = createErrorService()
        )
        dittoManager.ditto?.let { ditto ->
            // disable sync with v3 peers, required for DQL to work
            ditto.disableSyncWithV3()

            // Disable DQL strict mode
            // when set to false, collection definitions are no longer required. SELECT queries will return and display all fields by default.
            // https://docs.ditto.live/dql/strict-mode
            ditto.store.execute("ALTER SYSTEM SET DQL_STRICT_MODE = false")

            ditto.store.execute(
                "ALTER SYSTEM SET USER_COLLECTION_SYNC_SCOPES = :syncScopes",
                mapOf(
                    "syncScopes" to mapOf(
                        "tasks" to "LocalPeerOnly"
                    )
                )
            )
        }
        return dittoManager
    }

    private fun createErrorService() : ErrorService {
        return ErrorService()
    }

    private suspend fun cleanUpCollection(dittoManager: DittoManager) {
        try {
            dittoManager.stopSync()
            dittoManager.ditto?.store?.execute("EVICT FROM tasks")

            //remove the directory and all files
            val directoryPath = dittoManager.ditto?.persistenceDirectory
            directoryPath?.let { path ->
                val directory = File(path)
                if (directory.exists()) {
                    directory.listFiles()?.forEach { file ->
                        try {
                            file.delete()
                        } catch (e: Exception) {
                            Log.e("DittoManagerTests", "Failed to delete file: ${file.absolutePath}", e)
                        }
                    }
                    try {
                        directory.delete()
                    } catch (e: Exception) {
                        Log.e("DittoManagerTests", "Failed to delete directory: ${directory.absolutePath}", e)
                    }
                }
            }
        } catch (e: Exception) {
            Log.e("DittoManagerTests", "Failed to clean up collection", e)
        } finally {
            dittoManager.ditto = null
        }
    }

}

File: swift/TasksTests/DittoManagerTests.swift
import Testing
import Foundation
import DittoSwift

@testable import Tasks

@MainActor
struct DittoManagerTests {
    
    @Test
    func testDittoPopulateTaskCollection() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        
        // Assert
        try await dittoManager.populateTaskCollection()
        let tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 4, "Should have 4 initial tasks")
        
        // Verify specific tasks exist
        let expectedTitles = [
            "Buy groceries",
            "Clean the kitchen",
            "Schedule dentist appointment",
            "Pay bills"
        ]
        
        // Check each expected title exists exactly once
        for expectedTitle in expectedTitles {
            let matchingTasks = tasks.filter { $0.title == expectedTitle }
            #expect(matchingTasks.count == 1,
                    "Should find exactly one task with title '\(expectedTitle)', found \(matchingTasks.count)")
        }
        
        // Verify IDs are unique
        let uniqueIds = Set(tasks.map { $0._id })
        #expect(uniqueIds.count == tasks.count,
                "All task IDs should be unique")
        
        //clean up
        try await cleanUpCollection(dittoManager)
    }

    @Test
    func testDittoInsertTaskModel() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        var tasks = try await getTasksCollection(dittoManager)
        let newTask = createInitialTask()
        
        // Get initial count
        let initialCount = tasks.count
        
        // Act
        try await dittoManager.insertTaskModel(newTask)

        // get updated tasks
        tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == initialCount + 1, "Should have one more task after insertion")
        
        // Verify the new task was added correctly
        if let insertedTask = tasks.first(where: { $0._id == newTask._id }) {
            #expect(insertedTask.title == newTask.title, "Task title should match")
            #expect(insertedTask.done == newTask.done, "Task done status should match")
            #expect(insertedTask.deleted == newTask.deleted, "Task deleted status should match")
        } else {
            #expect(Bool(false), "New task should be found in tasks array")
        }
        
        //clean up
        try await cleanUpCollection(dittoManager)
    }
    
    @Test
    func testDittoUpdateTaskModel() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        var tasks = try await getTasksCollection(dittoManager)
        let initialTask = createInitialTask()
        
        // Add initial task
        try await dittoManager.insertTaskModel(initialTask)

        // Create updated version of task
        var updatedTask = initialTask
        updatedTask.title = "Updated Title"
        updatedTask.done = true
        
        // Act
        try await dittoManager.updateTaskModel(updatedTask)

        // Assert
        tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 1, "Should still have one task after update")
        if let resultTask = tasks.first(where: { $0._id == updatedTask._id }) {
            #expect(resultTask._id == initialTask._id, "Task ID should remain unchanged")
            #expect(resultTask.title == "Updated Title", "Task title should be updated")
            #expect(resultTask.done == true, "Task done status should be updated")
        } else {
            #expect(Bool(false), "Updated task should be found in tasks array")
        }
        
        //clean up
        try await cleanUpCollection(dittoManager)
    }
    
    @Test
    func testDittoUpdateTaskModel_WhenDeleted() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        var initialTask = createInitialTask()
        
        // Add initial task
        try await dittoManager.insertTaskModel(initialTask)
        var tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 1, "Should have 1 task after insert")
        
        // Act
        initialTask.deleted = true
        try await dittoManager.updateTaskModel(initialTask)

        // Assert
        tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 0, "Should have 0 task after update")
        
        //clean up
        try await cleanUpCollection(dittoManager)
    }
    
    @Test
    func testDittoToggleComplete() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        let initialTask = createInitialTask()
        
        // Add initial task
        try await dittoManager.insertTaskModel(initialTask)
        var tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 1, "Should have 1 task after insert")
        
        // Act - Toggle complete (false -> true)
        try await dittoManager.toggleComplete(task: initialTask)

        // Assert
        tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 1, "Should still have 1 task after update")
        if let resultTask = tasks.first(where: { $0._id == initialTask._id }) {
            #expect(resultTask._id == initialTask._id, "Task ID should remain unchanged")
            #expect(resultTask.title == initialTask.title, "Task title remain unchanged")
            #expect(resultTask.done == true, "Task should be marked as done after first toggle")
            
            // Assert a second toggle
            // Act - Toggle complete again (true -> false)
            try await dittoManager.toggleComplete(task: resultTask)
            tasks = try await getTasksCollection(dittoManager)
            if let finalTask = tasks.first(where: { $0._id == initialTask._id }) {
                #expect(finalTask.done == false, "Task should be marked as not done after second toggle")
                #expect(finalTask._id == initialTask._id, "Task ID should remain unchanged")
                #expect(finalTask.title == initialTask.title, "Task title should remain unchanged")
            } else {
                #expect(Bool(false), "Task should still exist after second toggle")
            }
            
        } else {
            #expect(Bool(false), "Updated task should be found in tasks array")
        }
    }
    
    @Test
    func testDittoDeleteTaskModel() async throws {
        // Arrange
        let dittoManager = try await createDittoManagerForTests()
        let initialTask = createInitialTask()
        
        // Add initial task
        try await dittoManager.insertTaskModel(initialTask)
        var tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 1, "Should have 1 task after insert")
        if let resultTask = tasks.first(where: { $0._id == initialTask._id }) {
            #expect(resultTask._id == initialTask._id, "Task ID should remain unchanged")
            #expect(resultTask.title == initialTask.title, "Task title remain unchanged")
        } else {
            #expect(Bool(false), "Inserted task should be found in tasks array")
        }
        
        // Act
        try await dittoManager.deleteTaskModel(initialTask)
        
        // Assert
        tasks = try await getTasksCollection(dittoManager)
        #expect(tasks.count == 0, "Task array should be 0 after deletion")
        #expect(!tasks.contains(where: { $0._id == initialTask._id }),
                "Deleted task should not be in array")
    }
    
    
    private func createInitialTask() -> Tasks.TaskModel {
        return Tasks.TaskModel(
            _id: UUID().uuidString,
            title: "Test Task",
            done: false,
            deleted: false
        )
    }
    
    private func getTasksCollection(_ dittoManager: Tasks.DittoManager) async throws -> [Tasks.TaskModel] {
        if let dittoInstance = dittoManager.ditto {
            let results = try await dittoInstance.store.execute(query: "SELECT * FROM tasks WHERE NOT deleted")
            let tasks = results.items.compactMap{
                Tasks.TaskModel($0.jsonString())
            }
            return tasks
        }
        return []
    }
    
    private func createDittoManagerForTests() async throws -> Tasks.DittoManager  {
        let dittoManager = Tasks.DittoManager()
        // setup logging
        DittoLogger.enabled = true
        DittoLogger.minimumLogLevel = .debug
        
        // Create unique test directory
        let testDirectoryPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("ditto_test_\(UUID().uuidString)")
        
        // Ensure directory exists
        try FileManager.default.createDirectory(
            at: testDirectoryPath,
            withIntermediateDirectories: true
        )
        do {
            // Initialize Ditto with test-specific directory
            dittoManager.ditto = Ditto(
                identity: .onlinePlayground(
                    appID: Env.DITTO_APP_ID,
                    token: Env.DITTO_PLAYGROUND_TOKEN,
                    enableDittoCloudSync: false,
                    customAuthURL: URL(string: Env.DITTO_AUTH_URL)
                ),
                persistenceDirectory: testDirectoryPath
            )
            try await dittoManager.ditto?.store.execute(
                query: "ALTER SYSTEM SET USER_COLLECTION_SYNC_SCOPES = :syncScopes",
                arguments: ["syncScopes":
                                [ "tasks": "LocalPeerOnly"]])
            
            try dittoManager.ditto?.disableSyncWithV3()
        } catch {
            // Clean up test directory if initialization fails
            try? FileManager.default.removeItem(at: testDirectoryPath)
        }
        return dittoManager
    }
    
    private func cleanUpCollection(_ dittoManager: Tasks.DittoManager) async throws {
        if let dittoInstance = dittoManager.ditto {
            dittoManager.subscription?.cancel()
            try await dittoInstance.store.execute(query: "EVICT FROM tasks")
            if (dittoInstance.isSyncActive) {
                dittoInstance.stopSync()
            }
            let directory = dittoInstance.persistenceDirectory
            try? FileManager.default.removeItem(at: directory)
        }
    }
}
